wallet-new/src/Cardano/Wallet/Kernel/Transactions.hs

```diff
@@ -0,0 +1,270 @@
+
+instance Buildable NewTransactionError where
+    build (CoinSelectionFailed hardErr) =
+        bprint ("CoinSelectionFailed " % build) hardErr
+
+instance Arbitrary NewTransactionError where
+    arbitrary = CoinSelectionFailed <$> arbitrary
+
+data PaymentError = PaymentNewTransactionError NewTransactionError
+                  | PaymentNewPendingError NewPendingError
+                  | PaymentSubmissionMaxAttemptsReached
+                  -- ^ When trying to send the newly-created transaction via
+                  -- 'newPending' and the submission layer, we hit the number
+                  -- of retries/max time allocated for the operation.
+
+instance Buildable PaymentError where
+    build (PaymentNewTransactionError txErr) =
+        bprint ("PaymentNewTransactionError " % build) txErr
+    build (PaymentNewPendingError npe) =
+        bprint ("PaymentNewPendingError " % build) npe
+    build PaymentSubmissionMaxAttemptsReached =
+        bprint "PaymentSubmissionMaxAttemptsReached"
+
+-- | Workhorse kernel function to perform a payment. It includes logic to
+-- stop trying to perform a payment if the payment would take more than 30
+-- seconds, as well as internally retrying up to 5 times to propagate the
+-- transaction via 'newPending'.
+pay :: ActiveWallet
+    -> IO Address
+    -- ^ A computation to generate change 'Address'es.
+    -> (Address -> Either CoinSelHardErr SafeSigner)
+    -> CoinSelectionOptions
+    -> HdAccountId
+    -- ^ The source @(root, account) from where the payment was
+    -- originated
+    -> NonEmpty (Address, Coin)
+    -- ^ The payees
+    -> IO (Either PaymentError Tx)
+pay activeWallet genChangeAddr signAddress opts accountId payees = do
+        retrying retryPolicy shouldRetry $ \rs -> do
+            tx <- newTransaction activeWallet genChangeAddr signAddress opts accountId payees
+            case tx of
+                 Left e      -> return (Left $ PaymentNewTransactionError e)
+                 Right txAux -> do
+                     succeeded <- newPending activeWallet accountId txAux
+                     case succeeded of
+                          Left e   -> do
+                              print (sformat build e)
+                              -- If the next retry would bring us to the
+                              -- end of our allowed retries, we fail with
+                              -- a proper error
+                              retriesLeft <- applyPolicy retryPolicy rs
+                              return . Left $ case retriesLeft of
+                                   Nothing ->
+                                       PaymentSubmissionMaxAttemptsReached
+                                   Just _  ->
+                                       PaymentNewPendingError e
+                          Right () -> return . Right . taTx $ txAux
+    where
+        retryPolicy :: RetryPolicyM IO
+        retryPolicy = constantDelay 5000000 <> limitRetries 6
```

Comment:
Fair point from @edsko : we might want to randomise the constantDelay to avoid threads simultaneously competing on the shared resources here -- perhaps we can leverage something like `fullJitterBackoff` from the `retry` package.